typing/typeclass.ml

```diff
@@ -1191,6 +1197,17 @@ and class_expr cl_num val_env met_env scl =
           cl_env = val_env;
           cl_attributes = scl.pcl_attributes;
          }
+  | Pcl_open (ovf, lid, e) ->
+      let used_slot = ref false in
+      let (path, new_val_env) = !Typecore.type_open ~used_slot ovf val_env scl.pcl_loc lid in
+      let (_path, new_met_env) = !Typecore.type_open ~used_slot ovf met_env scl.pcl_loc lid in
+      let cl = class_expr cl_num new_val_env new_met_env e in
+      rc {cl_desc = Tcl_open (ovf, path, lid, new_val_env, cl);
+          cl_loc = scl.pcl_loc;
+          cl_type = cl.cl_type;
+          cl_env = val_env;
```

Comment:
> My understanding was that the environment here is just there so that we can properly interpret identifiers.

Indeed, a quick look at translclass suggest that cl_env is mostly used on Tcl_ident node.  There is also `build_class_lets`, but it deals with Tcl_let directly (and return the inner env explicitly).  The remaining use is through `Translobj.oo_wrap/Translobj.oo_add_class`, which ends up defining the `top_env` in `transl_class` (and then passed to Env.diff).  And here I'm a bit lost.

> So any of the two environments is ok (the new one has just more information).
> If we want to do that, we should define a more precise semantics.

I think the least confusing semantics would be to store the val_env "around" the class expression.  This is similar to the semantics of exp_env, pat_env, mod_env.  Do you agree?

> At least when it was introduced, the goal was not to be able to use this information to compile source code.

I'm not sure what you are referring to, but in my understanding, this more is more about supporting external tools (documentation generators, IDE support, etc) so that they can do some lookups.  It's not really about allowing to compile source code fragment (although this might be useful as well).